crates/mpt/src/db/mod.rs

```diff
@@ -141,71 +129,85 @@ where
-        self.storage_roots
-            .insert(address, TrieNode::Blinded { commitment: trie_account.storage_root });
-
-        // If the account's code hash is not empty, fetch the bytecode and insert it into the cache.
-        let code = (trie_account.code_hash != KECCAK_EMPTY)
-            .then(|| {
-                let code = Bytecode::new_raw((self.code_by_hash_fetcher)(trie_account.code_hash)?);
-                Ok::<_, anyhow::Error>(code)
-            })
-            .transpose()?;
-
-        // Return a partial DB account. The storage and code are not loaded out-right, and are
-        // loaded optimistically in the `Database` + `DatabaseRef` trait implementations.
-        let mut info = AccountInfo {
-            balance: trie_account.balance,
-            nonce: trie_account.nonce,
-            code_hash: trie_account.code_hash,
-            code,
-        };
-        self.insert_contract(&mut info);
+        // Recompute the root hash of the trie.
+        self.root_node.blind();
 
-        Ok(DbAccount { info, ..Default::default() })
+        // Extract the new state root from the root node.
+        self.root_node.blinded_commitment().ok_or(anyhow!("State root node is not a blinded node"))
     }
 
-    /// Inserts the account's code into the cache.
-    ///
-    /// Accounts objects and code are stored separately in the cache, this will take the code from
-    /// the account and instead map it to the code hash.
+    /// Modifies the accounts in the storage trie with the given [BundleState] changeset.
     ///
     /// # Takes
-    /// - `account`: The account to insert the code for.
-    pub(crate) fn insert_contract(&mut self, account: &mut AccountInfo) {
-        if let Some(code) = &account.code {
-            if !code.is_empty() {
-                if account.code_hash == KECCAK_EMPTY {
-                    account.code_hash = code.hash_slow();
-                }
-                self.db.contracts.entry(account.code_hash).or_insert_with(|| code.clone());
+    /// - `bundle`: The [BundleState] changeset to apply to the trie DB.
+    ///
+    /// # Returns
+    /// - `Ok(())` if the accounts were successfully updated.
+    /// - `Err(_)` if the accounts could not be updated.
+    fn update_accounts(&mut self, bundle: &BundleState) -> Result<()> {
+        for (address, bundle_account) in bundle.state() {
+            let account_info =
+                bundle_account.account_info().ok_or(anyhow!("Account info not found"))?;
+            let mut trie_account = TrieAccount {
+                balance: account_info.balance,
+                nonce: account_info.nonce,
+                code_hash: account_info.code_hash,
+                ..Default::default()
+            };
+
+            // Update the account's storage root
+            let acc_storage_root = self
+                .storage_roots
+                .get_mut(address)
+                .ok_or(anyhow!("Storage root not found for account"))?;
```

Comment:
Should this insert on absense?

Reply:
For this one, no - if the account is included in the `BundleState`, it should already have a storage root cached. If not, we need to throw, because an account that wasn't loaded during execution is being altered.